### PR TITLE
Move corrupted checkpoint file

### LINKED ISSUE #86676: 1.18: panic in "pkg/kubelet/cm/qos_container_manager_linux.go"
**TL;DR:** we discovered the checkpoint file panic in the kubelet that was introduced recently.
see the comments bellow.

**What happened**:
created a 3 control-plane cluster from latest CI 1.17 using kubeadm.
initiated a upgraded from latest CI 1.17 to latest CI 1.18 using kubeadm.

pods such as api-server, etcd end up with stopped containers:
```
KUBECONFIG=/etc/kubernetes/admin.conf kubectl describe po -n kube-system etcd-kinder-upgrade-control-plane-1 | grep Events -A 10
Events:
  Type     Reason            Age                 From                                     Message
  ----     ------            ----                ----                                     -------
  Warning  DNSConfigForming  27m (x17 over 45m)  kubelet, kinder-upgrade-control-plane-1  Nameserver limits were exceeded, some nameservers have been omitted, the applied nameserver line is: 8.8.8.8 8.8.4.4 2001:4860:4860::8888
  Normal   Killing           26m                 kubelet, kinder-upgrade-control-plane-1  Stopping container etcd
```

the pod list gets stuck at this state, which is slightly incorrect as all containers running on the `-1` node are terminated:

```
root@kinder-upgrade-control-plane-1:/# KUBECONFIG=/etc/kubernetes/admin.conf kubectl get po -A   
NAMESPACE     NAME                                                     READY   STATUS        RESTARTS   AGE
kube-system   calico-kube-controllers-564b6667d7-j6qzw                 0/1     Running       0          42m
kube-system   calico-node-fg6mt                                        1/1     Running       0          39m
kube-system   calico-node-k94w9                                        1/1     Running       0          37m
kube-system   calico-node-nkfp6                                        0/1     Running       0          42m
kube-system   coredns-6955765f44-nvj5d                                 0/1     Running       0          42m
kube-system   coredns-6955765f44-trz6l                                 1/1     Running       0          24m
kube-system   etcd-kinder-upgrade-control-plane-1                      1/1     Running       0          42m
kube-system   etcd-kinder-upgrade-control-plane-2                      1/1     Running       0          39m
kube-system   etcd-kinder-upgrade-control-plane-3                      1/1     Running       0          37m
kube-system   kube-apiserver-kinder-upgrade-control-plane-1            1/1     Running       0          24m
kube-system   kube-apiserver-kinder-upgrade-control-plane-2            1/1     Running       0          39m
kube-system   kube-apiserver-kinder-upgrade-control-plane-3            1/1     Running       0          37m
kube-system   kube-controller-manager-kinder-upgrade-control-plane-1   1/1     Running       0          24m
kube-system   kube-controller-manager-kinder-upgrade-control-plane-2   1/1     Running       1          39m
kube-system   kube-controller-manager-kinder-upgrade-control-plane-3   1/1     Running       0          37m
kube-system   kube-proxy-cmssx                                         1/1     Running       0          39m
kube-system   kube-proxy-f9vvn                                         1/1     Terminating   0          42m
kube-system   kube-proxy-hqq22                                         1/1     Running       0          23m
kube-system   kube-scheduler-kinder-upgrade-control-plane-1            1/1     Running       0          24m
kube-system   kube-scheduler-kinder-upgrade-control-plane-2            1/1     Running       0          39m
kube-system   kube-scheduler-kinder-upgrade-control-plane-3            1/1     Running       0          37m
```

a panic is observed in the kubelet logs of the primary node:
```
$ journalctl -xeu kubelet

<snip>

Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]: goroutine 603 [running]:
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]: k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.logPanic(0x3a2dec0, 0x6d6df80)
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/
pkg/util/runtime/runtime.go:74 +0xa3
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]: k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/
pkg/util/runtime/runtime.go:48 +0x82
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]: panic(0x3a2dec0, 0x6d6df80)
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]:         /usr/local/go/src/runtime/panic.go:679 +0x1b2
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]: k8s.io/kubernetes/pkg/kubelet/cm.(*qosContainerManagerImpl).setCPUCgroupConfig(0xc000a8a0a0, 0xc000f14f38
, 0x428037b, 0xa)
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/cm/qos_containe
r_manager_linux.go:169 +0x3b
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]: k8s.io/kubernetes/pkg/kubelet/cm.(*qosContainerManagerImpl).UpdateCgroups(0xc000a8a0a0, 0x0, 0x0)
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/cm/qos_containe
r_manager_linux.go:285 +0x291
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]: k8s.io/kubernetes/pkg/kubelet/cm.(*containerManagerImpl).UpdateQOSCgroups(0xc000415200, 0xc000b7c800, 0xc
000d7e580)
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/cm/container_ma
nager_linux.go:561 +0x3a
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]: k8s.io/kubernetes/pkg/kubelet.(*Kubelet).killPod(0xc000c2e000, 0xc000b7c800, 0x0, 0xc000d15300, 0x0, 0x4,
 0x4)
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_pods.go
:832 +0x113
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]: k8s.io/kubernetes/pkg/kubelet.(*Kubelet).syncPod(0xc000c2e000, 0x0, 0xc000b7c800, 0x2, 0xc000d15300, 0x0,
 0xc000143320, 0xc000d3bbf0)
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1613
 +0x2493
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]: k8s.io/kubernetes/pkg/kubelet.(*podWorkers).managePodLoop.func1(0xc0007a2ea0, 0xc000213730, 0xc000f15e88,
 0xc0003271e0, 0x4a5ebc0)
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/pod_workers.go:
175 +0x25d
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]: k8s.io/kubernetes/pkg/kubelet.(*podWorkers).managePodLoop(0xc000213730, 0xc000959f20)
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/pod_workers.go:
184 +0x13f
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]: k8s.io/kubernetes/pkg/kubelet.(*podWorkers).UpdatePod.func1(0xc000213730, 0xc000959f20)
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/pod_workers.go:
222 +0x6c
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]: created by k8s.io/kubernetes/pkg/kubelet.(*podWorkers).UpdatePod
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/pod_workers.go:
220 +0x3d3
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]: E1227 21:41:30.920880   30265 runtime.go:78] Observed a panic: "invalid memory address or nil pointer der
eference" (runtime error: invalid memory address or nil pointer dereference)
Dec 27 21:41:30 kinder-upgrade-control-plane-1 kubelet[30265]: goroutine 497 [running]:
```

**What you expected to happen**:
the kubelet not killing containers.

**How to reproduce it (as minimally and precisely as possible)**:

it is possible with the tool `kinder` in the kubeadm repository, but it might take a while depending on your internet speed.

```
git clone git@github.com:kubernetes/kubeadm.git --depth=1
cd ./kubeadm/kinder
go build

docker pull kindest/base:v20190403-1ebf15f
./kinder build node-image-variant --base-image=kindest/base:v20190403-1ebf15f --image=kindest/node:test --with-init-artifacts=v1.17.1-beta.0.30+7d04da8a4a80f2 --with-upgrade-artifacts=v1.18.0-alpha.1.181+13133857aff48f --loglevel=debug
./kinder create cluster --name=kinder-upgrade --image=kindest/node:test --control-plane-nodes=3 --loglevel=debug
./kinder do kubeadm-init --name=kinder-upgrade --automatic-copy-certs --loglevel=debug --kubeadm-verbosity=6
./kinder do kubeadm-join --name=kinder-upgrade --automatic-copy-certs --loglevel=debug --kubeadm-verbosity=6
./kinder do kubeadm-upgrade --upgrade-version=v1.18.0-alpha.1.181+13133857aff48f --name=kinder-upgrade --loglevel=debug --kubeadm-verbosity=6
```
^ the last command (upgrade) should fail near upgrading the etcd member on the second CP node.

> Failed to get etcd status for https://172.17.0.3:2379: failed to dial endpoint https://172.17.0.3:2379 with maintenance client: context deadline exceeded

at this point it is possible to ctrl+c (or wait for timeout - 10minutes) and execute in the primary CP node:

```
docker exec -it  kinder-upgrade-control-plane-1 bash
kubectl --kubeconfig /etc/kubernetes/admin.conf kubectl describe po -n kube-system etcd-kinder-upgrade-control-plane-1
journalctl -xeu kubelet
crictl ps -a
```

to delete the cluster:
```
kinder delete cluster --name kinder-upgrade
```

**Anything else we need to know?**:
- the tool kinder is doing something wrong, it upgrades the primary CP node kubelet before upgrading the api server on the second CP node. this is incorrect upgrade order, which i will TODO/verify if it affects the panic behavior here. [1]
- the kinder runs containerd on the DIND nodes (based on kind).

this started failing after this DIFF:
https://github.com/kubernetes/kubernetes/compare/0b830f3d6...e5f0648c6

a PR that touches `pkg/kubelet/cm/container_manager_linux.go` in that diff is:
https://github.com/kubernetes/kubernetes/pull/84462

/sig node
/kind bug
cc @klueska @liggitt

i'm adding this priority:
/priority awaiting-more-evidence

but if it still fails once [1] is corrected, i will escalate the priority.

xref:
https://github.com/kubernetes/kubeadm/issues/1981

**Environment**:
- Kubernetes version (use `kubectl version`): see above.
- Cloud provider or hardware configuration: docker-in-docker
- OS (e.g: `cat /etc/os-release`): `Ubuntu 18.04.1 LTS` in the DIND nodes.
- Kernel (e.g. `uname -a`): `Linux kinder-upgrade-control-plane-1 4.13.0-41-generic #46-Ubuntu SMP Wed May 2 13:38:30 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools: kubeadm/kinder
- Network plugin and version (if this is a network-related bug): calico 3.8.2
- Others: NONE

### PR DESCRIPTION
**Pull request by tedyu**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
xref #86676 this PR moves corrupted checkpoint file.

This PR is orthogonal to the fix for any given case of checkpoint file corruption.
Suppose for some new case the operator follows the instruction and removes the corrupt checkpoint file, it may be difficult for developer to find the cause and provide fix in timely manner.

This PR sidelines the checkpoint file and allows more time for the developer to analyze the sidelined checkpoint file.

**Which issue(s) this PR fixes**:
xref #86676

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by neolit123** [comment]
> Fixes #86676 by moving corrupted checkpoint file.

i wouldn't qualify this as a fix to that issue.
i think there is a deeper problem at play.

/hold

**Comment by neolit123** [comment]
best to `xref #86676` instead of `fixes #86676`.

**Comment by mattjmcnaughton** [comment]
For those seeing this pr, but not the entire thread, I think https://github.com/kubernetes/kubernetes/pull/86689 fixes the underlying issue.

@tedyu is this pr still necessary given @klueska WIP diff that I linked above?

**Comment by tedyu** [comment]
This PR is orthogonal to the fix for any given case of checkpoint file corruption.
Suppose for some new case the operator follows the instruction and removes the corrupt checkpoint file, it may be difficult for developer to find the cause and provide fix in timely manner.

This PR sidelines the checkpoint file and allows more time for the developer to analyze the sidelined checkpoint file.

**Comment by mattjmcnaughton** [comment]
> This PR is orthogonal to the fix for any given case of checkpoint file corruption.
> Suppose for some new case the operator follows the instruction and removes the corrupt checkpoint file, it may be difficult for developer to find the cause and provide fix in timely manner.
> 
> This PR sidelines the checkpoint file and allows more time for the developer to analyze the sidelined checkpoint file.

I'll defer to @neolit123 here :) Whatever happens, thank you for always being so quick to jump in and tackle bugs :)

File: pkg/kubelet/cm/cpumanager/state/state_checkpoint.go [hunk 1]
```diff
@@ -54,10 +55,21 @@ func NewCheckpointState(stateDir, checkpointName, policyName string, initialCont
...
 
 	if err := stateCheckpoint.restoreState(); err != nil {
-		//lint:ignore ST1005 user-facing error message
-		return nil, fmt.Errorf("could not restore state from checkpoint: %v\n"+
-			"Please drain this node and delete the CPU manager checkpoint file %q before restarting Kubelet.",
-			err, path.Join(stateDir, checkpointName))
+		klog.Errorf("could not restore state from checkpoint: %v", err)
+		checkpointFile := path.Join(stateDir, checkpointName)
+		corruptpath := checkpointFile + ".corrupt"
+		if err := os.Rename(checkpointFile, corruptpath); err != nil {
```
Comment (liggitt) [inline pkg/kubelet/cm/cpumanager/state/state_checkpoint.go]: if a kubelet really did encounter this scenario, I'm not sure what the consequences are of discarding the invalid file and allowing the kubelet to continue without draining current pods.

**Comment by liggitt** [comment]
> This PR sidelines the checkpoint file and allows more time for the developer to analyze the sidelined checkpoint file.

is it ok to do that without draining the node?

**Comment by tedyu** [comment]
The draining of node should be performed first.

**Pull request closed without merging**